### ISSUE TITLE
Epick_d  and Has_filtered_predicates_tag

### ISSUE THREAD

**Issue by afabri** [CGAL/cgal#8256]
## Issue Details

The upcoming Frechet distance package works in any dimension.   [Here ](https://github.com/anusser/cgal/blob/a4687d4d3894e6058dbbbf809910ba7a2f71a698/Polyline_distance/include/CGAL/Polyline_distance/internal/curve.h#L71) we want to obtain `K::Has_filtered_predicates_tag::value` for the kernel, and we then want to access `K::Exact_kernel`.  
@mglisse  Can this be added to `Epick_d` without breaking its design?

**Comment by mglisse** [comment]
We can probably do something, but can you explain what you are trying to achieve there (otherwise I may answer the wrong question)?
I see that you mostly ignore the input kernel and do some computations always with intervals, and keep a way to extract vertices as rational points, but that doesn't really give me a high level image of the goal and strategy.

**Comment by afabri** [comment]
The input points are transformed in points with interval arithmetic.   In case of an uncertain exception we create a `Sqrt_extension` where we want to use the `Exact_kernel::FT`  if available.   
The table [here](https://github.com/anusser/cgal/blob/a4687d4d3894e6058dbbbf809910ba7a2f71a698/Polyline_distance/include/CGAL/Polyline_distance/internal/curve.h#L112) shows also in which case we make a copy of the input.

**Comment by lrineau** [comment]
@mglisse 

Would the following patch be acceptable?

```patch
diff --git a/NewKernel_d/include/CGAL/NewKernel_d/Cartesian_filter_K.h b/NewKernel_d/include/CGAL/NewKernel_d/Cartesian_filter_K.h
index ef44403f9f5..bc7d88deefa 100644
--- a/NewKernel_d/include/CGAL/NewKernel_d/Cartesian_filter_K.h
+++ b/NewKernel_d/include/CGAL/NewKernel_d/Cartesian_filter_K.h
@@ -58,11 +58,15 @@ struct Cartesian_filter_K : public Base_
     typedef Base_ Kernel_base;
     typedef AK_ AK;
     typedef EK_ EK;
+    typedef EK Exact_kernel;
     typedef typename Store_kernel<AK_>::reference_type AK_rt;
     AK_rt approximate_kernel()const{return sak.kernel();}
     typedef typename Store_kernel<EK_>::reference_type EK_rt;
     EK_rt exact_kernel()const{return sek.kernel();}
 
+    enum { Has_filtered_predicates = true };
+    typedef Boolean_tag<Has_filtered_predicates> Has_filtered_predicates_tag;
+
     // MSVC is too dumb to perform the empty base optimization.
     typedef std::bool_constant<
       internal::Do_not_store_kernel<Kernel_base>::value &&
```

**Comment by lrineau** [comment]
Plus the following:

```patch
diff --git a/NewKernel_d/test/NewKernel_d/Epick_d.cpp b/NewKernel_d/test/NewKernel_d/Epick_d.cpp
index 5693977869a..61fed477750 100644
--- a/NewKernel_d/test/NewKernel_d/Epick_d.cpp
+++ b/NewKernel_d/test/NewKernel_d/Epick_d.cpp
@@ -751,6 +751,9 @@ int main(){
   test2<Ker2>(); test2i<Ker2>();
   test3<Ker3>();
   test3<Kerd>();
+  static_assert(Ker2::Has_filtered_predicates_tag::value);
+  static_assert(Ker3::Has_filtered_predicates_tag::value);
+  static_assert(Kerd::Has_filtered_predicates_tag::value);
```

**Comment by afabri** [comment]
I won't make a dedicated pull request, but it will go in PR #8284

**Comment by mglisse** [comment]
> But... to be honest, I have not the impression that the patch will decrease the readability of the code! 😃

If we want the probability of a cleanup to be non-zero, we should
* avoid adding more things to clean up
* mark clearly the things that should **not** be cleaned up (this was my point here)

(I completely agree the code is not pretty)

> In my editor (VS Code), the TODO and FIXME are highlighted, there are.. a lot!

Comments (whether they are marked TODO/FIXME or not) are a completely different thing. They usually contain important information. And yes, there are a lot of things that I never did because they were not priorities.

> That is probably to construct kind of "filtered constructions", where the constructed objects are sometimes reconstructed using an exact kernel, if needed. I do not know the details.

I had an Epack_d branch (it should still exist) to do filtered constructions in dD, but it didn't offer the option for each functor to use a different exact type (rational, sqrt-extension, etc).

> See this example of a valid usage:

It looks like it would have been much simpler to add Side_of_mixed_cell_3 to the kernel.

**Comment by afabri** [comment]
In fact I also need a function object for constructing the filtered type and the exact type, In `Epick`  this are the classes `C2F`  and `C2E`.

**Comment by lrineau** [comment]
> > But... to be honest, I have not the impression that the patch will decrease the readability of the code! 😃
> 
> If we want the probability of a cleanup to be non-zero, we should
> 
>     * avoid adding more things to clean up
> 
>     * mark clearly the things that should **not** be cleaned up (this was my point here)

You are right. And the comments are indeed very important for that (both TODO/FIXME and notes about why something has been added).

> > That is probably to construct kind of "filtered constructions", where the constructed objects are sometimes reconstructed using an exact kernel, if needed. I do not know the details.
> 
> I had an Epack_d branch (it should still exist) to do filtered constructions in dD, but it didn't offer the option for each functor to use a different exact type (rational, sqrt-extension, etc).

I was not able to retrieve the branch? Do you still have it? What is its name?

**Comment by mglisse** [comment]
> > I had an Epack_d branch (it should still exist) to do filtered constructions in dD, but it didn't offer the option for each functor to use a different exact type (rational, sqrt-extension, etc).
> 
> I was not able to retrieve the branch? Do you still have it? What is its name?

https://github.com/mglisse/cgal/tree/NewKernel_d-Epack-glisse
